### PR TITLE
Memory Inspector: UI and UX Updates and New Features

### PR DESCRIPTION
**Pull request by colin-grant-work**
### What it does

This PR introduces substantial changes to the UI of the memory viewer widget, building on Simon Marchi's work on colorizing variables. It adds useful functionality for inspecting memory during debugging including: dynamic updating of memory, multiple memory views, view comparison (diffing), register viewing.

---

#### Multiple Views

We have added the ability to open multiple memory views. This allows users to compare different regions of memory or to examine the same memory region of memory at different points in time.

https://user-images.githubusercontent.com/62660806/110188045-054cd600-7de0-11eb-9d7c-684ffe08b3da.mp4

#### Diff View

In addition to the standard memory viewer, we have added a diffing view that allows users to compare the content of different memory views in-line.

https://user-images.githubusercontent.com/62660806/110188129-49d87180-7de0-11eb-9c6e-1fb537f668d5.mp4

#### Dynamic Updating and Locking

Memory views will now update as users step through code, and changes to the visible memory region will be highlighted. If users would like to opt out of automatic updates - for example to preserve a memory state for comparison to a later state - they can 'lock' the view by clicking a lock icon.

https://user-images.githubusercontent.com/62660714/110826053-4f0f4380-825a-11eb-8bb9-2b4e23971560.mp4

https://user-images.githubusercontent.com/62660714/110826066-533b6100-825a-11eb-9c0c-c4e7b9181de5.mp4

#### Entry points

On the master branch, in order to examine a particular memory region, users must manually open the memory view and configure its view parameters when stopped at a breakpoint. This PR adds context-menu commands to open a memory view or register view at the location of a variable visible in the Debug Variables widget while also allowing manually address specification.

https://user-images.githubusercontent.com/62660714/110826152-677f5e00-825a-11eb-8695-9e00bcfe42ad.mp4

https://user-images.githubusercontent.com/62660714/110826935-2fc4e600-825b-11eb-9bcf-39f5152606f9.mp4

#### Register View

The Debug Adapter Protocol supports the reporting of register contents as a variable scope, although the CDT-GDB adapter doesn't currently provide registers. We have added a memory view dedicated to showing register contents, since the layout is quite different from the standard memory view.

---

### Setup

> Prerequisites: You should have [GDB](https://www.gnu.org/software/gdb/) set up on your system and have a working installation of [NodeJS](https://nodejs.org/en/) >= v.12 and < v.13.

1. Clone the repo and checkout this PR (`git fetch origin pull/119/head:memory-inspector && git checkout memory-inspector`)
2. Build the application (`npx yarn`)
3. If you intend to use `electron`, you will need to run `npx yarn rebuild:electron`.
4. Start the application (`npx yarn start:browser` or `npx yarn start:electron`)
5. Open the example workspace (File > Open Workspace... >  select `<repository>/examples/cpp-debug-workspace`)
6. Run the compilation task: (Terminal > Run Task... > select `compile example`)
7. Open `a.cpp` and add a break point somewhere in `main`.
8. Open the debugger using the image icon.
9. Launch a debug session using configuration `test-cpp` - it should already be selected in the dropdown. image
10. Follow some of the steps below to explore the functionality of the 

### How to test

* **Getting started:**
  - Open the `examples/cpp-debug-workspace` workspace and run the command `Task: Run` and run the compilation task.
  - Set a breakpoint in `a.cpp`
  - open the debug widget and run the debug configuration from the `launch.json`
* **Accessing the Memory View (2 ways)**:
  - Via the Debug Widget
    - Open the variables > locals pane in the Debug Widget
    - Right click on a variable name (or struct member)
    - Click 'Show variable in memory view'
    - You should see a memory widget whose address field is set to `&<what you clicked>` and it should be populated with colorful memory.
  - From the Memory View
    - Click on the Memory View icon or run the command `Toggle Memory View`
    - You should get a blank widget.
    - You can enter an address as a number or `&<some variable>` to get memory at that location.
* **Freezing Memory (Before you do anything else)**:
  - Click the lock icon at the top of the widget.
* **Starting up a second widget:**
  - Get a new widget with the same display by following the same steps as above (to get a clean widget, click the `hex` icon in the Memory Widget Tabbar)
  - You should see a diff compare icon appear in the tabbar for the main widget.
     - If you click it, a small widget at the bottom of the Memory Widget pane should appear that gives you options to select your widgets for comparison. Just ignore it… for now 😊
  - Then, in the second widget:
    - Change the display settings
      - Click the `Settings` cog wheel at the top
      - Manipulate the various fields that appear.
      - The content of memory you see shouldn&#39;t change, but you can change its layout
    - Change the Title
      - Hover over the large label for the widget.
      - You should see a pencil icon appear
      - Click anywhere over the title or pencil.
      - The title should transform into an input box.
      - Input something new and memorable.
      - Hitting `enter` should save, `escape` should cancel the edit.
      - If you save, the tab for the widget should also update.
  - Move the widget around/change the widget location
    - Move the widget to a different area of the application by dragging the tab
* **Dynamic features:**
  - Responding to Value Setting
    - In the Debug Widget, find a variable that is visible in the memory widget.
    - Right click on that variable and select &#39;set value&#39;. Set it to a different value than it currently has.
    - The memory widget should update to reflect the value you set, and you should see a highlighted memory field indicating a change.
  - Responding to changes during steps
    - Find some code that exercises a variable that is visible in the memory view.
    - Step through that code.
    - When changes are made relative to what is shown in the memory view, you should see a highlight.
  - Hovering
    - In a standard widget, you should be able to hover over a memory cell and see various data:
      - Hex, binary, octal, decimal, and Unicode value should appear for every cell.
      - If the cell changed during a step, you should see the previous value.
      - If you&#39;ve typed a new value into the cell, you should see the current value of the byte.
  - Diffing
    - Good scenario:
      - Now that the first widget you opened up and locked and the second widget are different, click on the `Go` button in the comparison area.
      - You can diff any arbitrary memory regions by changing the offset from the start of the widget.
      - A widget should appear that shows the memory region you&#39;ve been examining, with familiar diff highlights.
      - You have the same display formatting options as in a normal widget.
      - You can view ASCII, and it should be diffed.
    - Bad scenarios
      - If you have no memory loaded in one of the widgets selected for diffing, you should get a toast saying you need to have memory in both widgets.
      - If you select two widgets with non-overlapping memory areas, you should get a toast saying that you can&#39;t diff non-overlapping memory
    - Other Diff Select Widget Features
      - It should be impossible to try to diff the same widget against itself.
        - If you open the first select input and choose the same value as in the second select, the second select should change to some other value.
        - If you open the second select input, you shouldn&#39;t see the value of the first input as an option.
      - It should update with changes to widget title.
        - Change the title of one of the widgets you see selected.
        - The selection should update
- Cleanup
  - Put your memory and diff widgets wherever you want in the shell.
  - Restart the application.
  - They should all be gone.

---

**Additional features to observe:**

  - **Memory Pagination** : When you are viewing memory, you have the option to &quot;Load X more bytes&quot; at the top/bottom of the table. Clicking on this text will perform the action, you can also use the number-select dropdown to load more/less memory. The input fields for &quot;offset&quot; and &quot;length&quot; should respond accordingly.
  - **Theming:** light/dark/high contrast
  - **Keyboard accessibility:** are all inputs tabbable, do arrow keys/enter/escape work as expected
  - **Dynamic icons** : widget icon will toggle between &quot;0x&quot; and &quot;lock&quot; when widget state changes
  - **Static/unpopulated views** : If you open the memory view without any data loaded, are the empty views aesthetically pleasing?
  - **Resizing** : does the widget respond (mostly) well to reasonable window resize changes?

---
To think about:
- [ ] The current memory provider works with the cdt-gdb adapter, but most of the functionality is available via standard requests of the debug adapter protocol, so it might be desirable to implement a plain DAP memory provider, perhaps with a service to match sessions to the memory provider best-suited to handle them.
- [ ] Only the memory provider needs to know about how the debug adapter works - it's the only part of the material in src/browser that knows anything about the CDT-GDB adapter. The remaining frontend code is debugger-agnostic, as long as it can get memory from that provider. That means that the CPP part of the extension name is limited to the registering the GDB debug types on the backend. What does that mean for the nature of the repository?

### Review checklist

- [ ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/cpp-debug/src/browser/memory-provider/memory-provider.ts
Comment (colin-grant-work) [inline packages/cpp-debug/src/browser/memory-provider/memory-provider.ts:22]: @thegecko, this is the interface between the debugger backend and the frontend. The current implementation knows it's interacting with GDB because the CDT-GDB adapter doesn't implement memory reading and writing as prescribed by the debug adapter protocol, but it should be relatively straightforward to swap it out.

It might be desirable to implement something like a `MemoryProviderService` that gathers `MemoryProvider` implementations and matches a given debug session with the memory provider designed for it. Let us know if that sounds useful, and we'll be happy to add it.

**Comment by JonasHelming** [comment]
This looks great! Are there plans to merge this PR? This would make it so much easier to adopt the great work you did!

**Comment by JonasHelming** [comment]
@MatthewKhouzam

**Comment by colin-grant-work** [comment]
@JonasHelming, we will plan on merging next week or the following. The one outstanding desideratum is that the memory reading be reimplemented to use the debug adapter protocol's prescribed method, rather than a custom request. That functionality has been implemented for the CDT-GDB adapter (https://github.com/eclipse-cdt/cdt-gdb-adapter/pull/180)[here], and once the changes have been confirmed to work for this extension, we will rewrite the code in this PR to use the new method.

File: packages/cpp-debug/src/browser/memory-provider/memory-provider.ts
Comment (colin-grant-work) [inline packages/cpp-debug/src/browser/memory-provider/memory-provider.ts:99]: The write request is now in the DAP, so it should be possible to get it into the adapter.

**Comment by JonasHelming** [comment]
@colin-grant-work when do you expect to merge? I don not intend to push, it is just for planning our adoption use case.

**Comment by colin-grant-work** [comment]
@JonasHelming, at the moment, I am waiting on a [new release of CDT-GDB](https://github.com/eclipse-cdt/cdt-gdb-adapter/pull/180) to convert the calls to read memory to DAP-compatible requests. It seems that [hasn't yet](https://github.com/theia-ide/cdt-gdb-vscode/releases/) happened. If you'd like to see it merged earlier, I could write a DAP-compatible memory provider, but leave it unbound for now until a DAP-compatible memory request is available from CDT-GDB.

**Comment by marcdumais-work** [comment]
@paul-marechal Considering that you have since become a committer on the CDT project, maybe we can think about making releases [upstream](https://github.com/eclipse-cdt/cdt-gdb-vscode), instead of on our fork. I see there are still no release tags on the upstream repo, so I guess no official releases yet? 

Alternatively, as I recall, the way we currently consume the DA from the `cdt-gdb-vscode` extension is quite twisted - maybe it would be better to work on making the DA more directly consumable, and then that extension is not longer relevant to us?

**Comment by thegecko** [comment]
> Alternatively, as I recall, the way we currently consume the DA from the cdt-gdb-vscode extension is quite twisted - maybe it would be better to work on making the DA more directly consumable, and then that extension is not longer relevant to us?

https://open-vsx.org/extension/eclipse-cdt/cdt-gdb-vscode ??

**Comment by marcdumais-work** [comment]
> https://open-vsx.org/extension/eclipse-cdt/cdt-gdb-vscode ??

Good catch - the extension is published on `open-vsx`. I made that happen with this PR:  https://github.com/open-vsx/publish-extensions/pull/285 . I am not sure, but I think the bot may publish a new version automatically, if a new release tag is added to the repo.

**Comment by colin-grant-work** [comment]
At the moment it actually downloads from the GitHub releases:

https://github.com/eclipse-theia/theia-cpp-extensions/blob/0a5deb4382bb9a8335062b8165ef80707eba226c/packages/cpp-debug/package.json#L56-L58

File: packages/cpp-debug/src/browser/wrapper-widgets/memory-layout-widget.tsx
Comment (federicobozzini) [inline packages/cpp-debug/src/browser/wrapper-widgets/memory-layout-widget.tsx:90]: It looks like the `super` call is missing here and it's not possible to properly dispose of this widget. Can we add this?
```suggestion
        this.super.dispose();
    }
```

**Review by federicobozzini — COMMENTED** [review]
@colin-grant-work I've added one small suggestion before this is merged. Can you please take a look at it?

**Comment by JonasHelming** [comment]
Will this be merged soon? I do not want to push, but I need to adapt some project planning based on this, so it would be useful for me to have an ETA. Thank you in advance! :-)

**Comment by colin-grant-work** [comment]
@JonasHelming, as of yesterday, a [new release of CDT-GDB-VSCode](https://open-vsx.org/extension/eclipse-cdt/cdt-gdb-vscode) is available on Open VSX, so I can now make the changes to use the DAP and then merge this PR. Apologies for the delays!

**Comment by colin-grant-work** [comment]
@JonasHelming, I have made the necessary adjustments to use the DAP - just need to bring it up to date with current Theia and it will be good to go. Plan to take care of that tomorrow morning.

**Comment by JonasHelming** [comment]
sounds great, thank you very much for the update!

**Comment by colin-grant-work** [comment]
@JonasHelming, @federicobozzini, @thegecko, @kenneth-marut-work, @paul-marechal, I have pushed up some significant changes to this PR.
 - I have changed the method by which the CDT-GDB-VSCode plugin is consumed: now it is downloaded as a plugin, and the backend code that registered its functionality using Theia API's has been removed.
 - I have changed the `MemoryProvider` code to use DAP methods.
 - I have tried to trim the dependencies listed and dev-packages included from the Theia repo to the minimum necessary to build and run the application and the various NPM scripts defined in the `package.json` files.
 - I've committed the `yarn.lock`, though it can be omitted if people prefer not to have it - I didn't realize it hadn't been committed previously.
 
If you're willing, please give things a run through to ensure that everything is working to your satisfaction, and if it is, I'll merge this ASAP.

File: packages/cpp-debug/src/browser/memory-provider/memory-provider.ts
Comment (federicobozzini) [inline packages/cpp-debug/src/browser/memory-provider/memory-provider.ts:88]: The `address` property is described by the [DAP docs](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_ReadMemory) as 

> The address of the first byte of data returned.
> Treated as a hex value if prefixed with '0x', or as a decimal value otherwise.

Do you think it might make sense to also manage the case the address is represented as a decimal number to ensure full compatibility?

File: packages/cpp-debug/src/browser/memory-provider/memory-provider.ts
Comment (federicobozzini) [inline packages/cpp-debug/src/browser/memory-provider/memory-provider.ts:87]: The [DAP docs](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_ReadMemory) describes the `data` property as

> The bytes read from memory, encoded using base64.

while `hex2bytes` expects a hex encoded string. Isn't there some mismatch here?

**Review by federicobozzini — COMMENTED** [review]
@colin-grant-work I've added a couple of comments again.

**Comment by colin-grant-work** [comment]
@federicobozzini, I've pushed up a base-64 patch on this PR, but I propose that we merge it without those changes, since GDB is currently the main (only?) adapter that we here have an interest in, and it's sending its results in hex. I'll make a separate PR for the base 64 changes, and then there's less urgency on the long pipeline from CDT-GDB--> CDT-GDB-VSCode --> here. Does that sound reasonable?

cc: @JonasHelming, @thegecko

**Comment by colin-grant-work** [comment]
Unless there are any objections, I'll merge this on Monday.

**Comment by colin-grant-work** [comment]
@paul-marechal, would you mind approving this when you have a chance? Thanks!

**Review by paul-marechal — APPROVED** [review]
I actually can't test it locally as I don't have the right environment.

Looking at the code it looks good.

Looking at the animated captures, the new widgets looks much better than what we have.

@JonasHelming @thegecko @federicobozzini we plan on deprecating `@theia/cpp-debug` and rename this package something like `@theia/debug-memory-inspector`. The reasoning is that `cpp-debug` used to be about providing C/C++ debugging capabilities, but this is not the case anymore. We'll merge this PR for now, but keep in mind that we need a new package for it with a new more appropriate name, feel free to give suggestions.

**Pull request merged**
